File: Code/Node-Backend/src/services/billing.service.ts
import { randomUUID } from "crypto";
import { Product, Customer } from "../models/product.model";
import {
  Invoice,
  LineItem,
  CreateInvoiceInput,
} from "../models/invoice.model";
import { Mutex } from "./mutex";

const DEFAULT_TAX_RATE = 0.18; // 18% GST unless a product overrides it

class ProductNotFoundError extends Error {
  constructor(code: string) {
    super(`Product not found: ${code}`);
    this.name = "ProductNotFoundError";
  }
}

class InvoiceNotFoundError extends Error {
  constructor(id: string) {
    super(`Invoice not found: ${id}`);
    this.name = "InvoiceNotFoundError";
  }
}

class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

export { ProductNotFoundError, InvoiceNotFoundError, ValidationError };

class BillingService {
  private products = new Map<string, Product>();
  private customers = new Map<string, Customer>();
  private invoices = new Map<string, Invoice>();
  private billNumberCounter = 1000;
  private readonly mutex = new Mutex();

  constructor() {
    this.seed();
  }

  private seed(): void {
    const products: Product[] = [
      { code: "P001", name: "Basmati Rice 5kg", price: 450, taxRate: 0.05 },
      { code: "P002", name: "Sunflower Oil 1L", price: 180, taxRate: 0.18 },
      { code: "P003", name: "Toothpaste 100g", price: 55, taxRate: 0.18 },
      { code: "P004", name: "Notebook 200pg", price: 40, taxRate: 0.12 },
    ];
    for (const p of products) this.products.set(p.code, p);

    const customer: Customer = {
      id: "C1001",
      name: "Ramesh Kumar",
      specialPricing: { P002: 165 }, // negotiated price on sunflower oil
    };
    this.customers.set(customer.id, customer);

    console.log(
      `Seeded ${this.products.size} products and ${this.customers.size} customer(s).`
    );
  }

  getProduct(code: string): Product {
    const product = this.products.get(code.toUpperCase());
    if (!product) throw new ProductNotFoundError(code);
    return product;
  }

  getInvoice(id: string): Invoice {
    const invoice = this.invoices.get(id);
    if (!invoice) throw new InvoiceNotFoundError(id);
    return invoice;
  }

  listInvoices(dateFilter?: string): Invoice[] {
    const all = Array.from(this.invoices.values()).sort((a, b) =>
      a.createdAt < b.createdAt ? 1 : -1
    );
    if (!dateFilter) return all;
    return all.filter((inv) => inv.createdAt.startsWith(dateFilter));
  }

  async createInvoice(input: CreateInvoiceInput): Promise<Invoice> {
    if (!input.items || input.items.length === 0) {
      throw new ValidationError("At least one line item is required");
    }
    for (const item of input.items) {
      if (!item.productCode || item.quantity <= 0) {
        throw new ValidationError(
          "Each item needs a productCode and a positive quantity"
        );
      }
    }

    const customer = input.customerId
      ? this.customers.get(input.customerId)
      : undefined;
    if (input.customerId && !customer) {
      throw new ValidationError(`Unknown customerId: ${input.customerId}`);
    }

    // Resolve pricing up front so a bad product code fails before we ever
    // touch the shared bill-number counter.
    const lineItems: LineItem[] = input.items.map((item) => {
      const product = this.getProduct(item.productCode);
      const unitPrice =
        customer?.specialPricing?.[product.code] ?? product.price;
      const taxRate = product.taxRate ?? DEFAULT_TAX_RATE;
      const lineSubtotal = round2(unitPrice * item.quantity);
      const lineTax = round2(lineSubtotal * taxRate);
      return {
        productCode: product.code,
        quantity: item.quantity,
        unitPrice,
        taxRate,
        lineSubtotal,
        lineTax,
        lineTotal: round2(lineSubtotal + lineTax),
      };
    });

    const subtotal = round2(lineItems.reduce((s, i) => s + i.lineSubtotal, 0));
    const tax = round2(lineItems.reduce((s, i) => s + i.lineTax, 0));
    const total = round2(subtotal + tax);

    // Only the bill-number assignment + store write is serialized; pricing
    // above can run concurrently for different requests.
    return this.mutex.runExclusive(() => {
      const billNumber = `INV-${String(this.billNumberCounter++).padStart(
        6,
        "0"
      )}`;
      const invoice: Invoice = {
        id: randomUUID(),
        billNumber,
        customerId: customer?.id,
        items: lineItems,
        subtotal,
        tax,
        total,
        status: "OPEN",
        createdAt: new Date().toISOString(),
      };
      this.invoices.set(invoice.id, invoice);
      return invoice;
    });
  }

  async cancelInvoice(id: string): Promise<Invoice> {
    return this.mutex.runExclusive(() => {
      const invoice = this.getInvoice(id);
      // Idempotent: cancelling an already-cancelled bill just returns it.
      // The bill number is never reassigned or reused (GST compliance).
      invoice.status = "CANCELLED";
      return invoice;
    });
  }
}

function round2(n: number): number {
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

export const billingService = new BillingService();

File: Code/SpringBoot-Backend/src/main/java/com/aadhirai/billing/service/BillingService.java
package com.aadhirai.billing.service;

import com.aadhirai.billing.controller.CreateInvoiceRequest;
import com.aadhirai.billing.model.*;
import com.aadhirai.billing.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.UUID;

@Service
public class BillingService {

    private static final BigDecimal DEFAULT_TAX_RATE = new BigDecimal("0.18"); // 18% GST unless overridden

    private final ProductRepository productRepository;
    private final CustomerRepository customerRepository;
    private final InvoiceRepository invoiceRepository;
    private final BillNumberSequenceRepository billNumberSequenceRepository;

    public BillingService(ProductRepository productRepository, CustomerRepository customerRepository,
                           InvoiceRepository invoiceRepository, BillNumberSequenceRepository billNumberSequenceRepository) {
        this.productRepository = productRepository;
        this.customerRepository = customerRepository;
        this.invoiceRepository = invoiceRepository;
        this.billNumberSequenceRepository = billNumberSequenceRepository;
    }

    public Product getProduct(String code) {
        return productRepository.findById(code.toUpperCase())
                .orElseThrow(() -> new ProductNotFoundException(code));
    }

    public Invoice getInvoice(UUID id) {
        return invoiceRepository.findById(id).orElseThrow(() -> new InvoiceNotFoundException(id.toString()));
    }

    public List<Invoice> listInvoices(String date) {
        if (date == null || date.isBlank()) {
            return invoiceRepository.findAllByOrderByCreatedAtDesc();
        }
        Instant from = Instant.parse(date + "T00:00:00Z");
        Instant to = from.plus(1, ChronoUnit.DAYS);
        return invoiceRepository.findByCreatedAtBetweenOrderByCreatedAtDesc(from, to);
    }

    /**
     * Pricing is resolved before the transaction that touches the shared
     * bill-number counter, so a bad product code fails fast without ever
     * acquiring the counter's row lock.
     */
    @Transactional
    public Invoice createInvoice(CreateInvoiceRequest request) {
        if (request.getItems() == null || request.getItems().isEmpty()) {
            throw new BillingValidationException("At least one line item is required");
        }
        for (CreateInvoiceRequest.Item item : request.getItems()) {
            if (item.getProductCode() == null || item.getProductCode().isBlank() || item.getQuantity() <= 0) {
                throw new BillingValidationException("Each item needs a productCode and a positive quantity");
            }
        }

        Customer customer = null;
        if (request.getCustomerId() != null && !request.getCustomerId().isBlank()) {
            customer = customerRepository.findById(request.getCustomerId())
                    .orElseThrow(() -> new BillingValidationException("Unknown customerId: " + request.getCustomerId()));
        }

        BigDecimal subtotal = BigDecimal.ZERO;
        BigDecimal tax = BigDecimal.ZERO;
        List<LineItem> lineItems = new java.util.ArrayList<>();

        for (CreateInvoiceRequest.Item item : request.getItems()) {
            Product product = getProduct(item.getProductCode());
            BigDecimal unitPrice = product.getPrice();
            if (customer != null && customer.getSpecialPricing().containsKey(product.getCode())) {
                unitPrice = customer.getSpecialPricing().get(product.getCode());
            }
            BigDecimal taxRate = product.getTaxRate() != null ? product.getTaxRate() : DEFAULT_TAX_RATE;
            BigDecimal lineSubtotal = unitPrice.multiply(BigDecimal.valueOf(item.getQuantity())).setScale(2, RoundingMode.HALF_UP);
            BigDecimal lineTax = lineSubtotal.multiply(taxRate).setScale(2, RoundingMode.HALF_UP);
            BigDecimal lineTotal = lineSubtotal.add(lineTax);

            subtotal = subtotal.add(lineSubtotal);
            tax = tax.add(lineTax);
            lineItems.add(new LineItem(product.getCode(), item.getQuantity(), unitPrice, taxRate, lineSubtotal, lineTax, lineTotal));
        }
        subtotal = subtotal.setScale(2, RoundingMode.HALF_UP);
        tax = tax.setScale(2, RoundingMode.HALF_UP);
        BigDecimal total = subtotal.add(tax);

        String billNumber = nextBillNumber();
        Invoice invoice = new Invoice(billNumber, customer != null ? customer.getId() : null, subtotal, tax, total);
        lineItems.forEach(invoice::addItem);
        return invoiceRepository.save(invoice);
    }

    @Transactional
    public Invoice cancelInvoice(UUID id) {
        Invoice invoice = getInvoice(id);
        // Idempotent: cancelling an already-cancelled bill just returns it.
        // The bill number is never reassigned or reused (GST compliance).
        invoice.setStatus(Invoice.Status.CANCELLED);
        return invoiceRepository.save(invoice);
    }

    private String nextBillNumber() {
        BillNumberSequence seq = billNumberSequenceRepository.findForUpdate()
                .orElseGet(() -> billNumberSequenceRepository.save(new BillNumberSequence()));
        seq.setLastValue(seq.getLastValue() + 1);
        billNumberSequenceRepository.save(seq);
        return String.format("INV-%06d", seq.getLastValue());
    }
}
